graphql_compiler/compiler/emit_sql.py

```diff
@@ -673,50 +716,71 @@ class ColumnRouter:
 class CompilationState(object):
     """Mutable class used to keep track of state while emitting a sql query."""
 
-    def __init__(self, sql_schema_info, ir):
+    def __init__(self, sql_schema_info: SQLAlchemySchemaInfo, ir: IrAndMetadata):
         """Initialize a CompilationState, setting the current location at the root of the query."""
         # Immutable metadata
-        self._sql_schema_info = sql_schema_info
-        self._ir = ir
-        self._used_columns = _find_columns_used_outside_folds(sql_schema_info, ir)
+        self._sql_schema_info: SQLAlchemySchemaInfo = sql_schema_info
+        self._ir: IrAndMetadata = ir
+        self._used_columns: Dict[VertexPath, Set[str]] = _find_columns_used_outside_folds(
+            sql_schema_info, ir
+        )
         # mapping fold paths to FoldScopeLocations with field information
-        self._all_folded_fields = _find_folded_fields(ir)
+        self._all_folded_fields: Dict[FoldPath, Set[FoldScopeLocation]] = _find_folded_fields(ir)
 
         # Current query location state. Only mutable by calling _relocate.
-        self._current_location = None  # the current location in the query. None means global.
-        self._current_alias = None  # a sqlalchemy table Alias at the current location
+        self._current_location: Optional[
+            BaseLocation
+        ] = None  # The current location in the query. None means global.
+        self._current_alias: Optional[
+            Alias
+        ] = None  # SQLAlchemy table Alias at the current location.
+
+        # Current folded subquery state.
+        self._current_fold: Optional[
+            SQLFoldObject
+        ] = None  # SQLFoldObject to collect fold info and create folded subqueries.
 
         # Current folded subquery state.
         self._current_fold = None  # SQLFoldObject to collect fold info and guide output query
 
         # Dict mapping (some_location.query_path, fold_scope_location.fold_path) tuples to
-        # corresponding table _Aliases. some_location is either self._current_location
-        # or the base location of an open FoldScopeLocation.
+        # corresponding table Aliases. some_location is either self._current_location
+        # or the base location of an open FoldScopeLocation. For Locations, the second argument of
+        # the tuple will be None.
         # Note: for tables with an _x_count column, that column will always
         # be named "fold_output__x_count".
-        self._aliases = {}
+        self._aliases: Dict[Tuple[QueryPath, Optional[FoldPath]], Alias] = {}
```

Comment:
This is a bit more complicated. The value is sometimes a `ColumnRouter`. I'm thinking about making a common base class for both, but we could leave it untyped or have a union type meanwhile.

Reply:
made it a union for now